File: mid/Dynamic Programming/198. House Robber II (Space-Opt DP).cpp
// https://leetcode.com/problems/house-robber-ii/solutions/59921/9-lines-0ms-o-1-space-c-solution/
// Time complexity : O(n)
// Space complexity : O(1)
// Space-Optimzed DP

class Solution {
public:
    int rob(vector<int>& nums) {
        // Robbing the first house and skipping the last house. (Rob houses 0 to n - 2;)
        // Robbing the last house and skipping the first house. (Rob houses 1 to n - 1)
        int n = nums.size(); 
        if (n < 2) return n ? nums[0] : 0;
        return max(helper(nums, 0, n - 2), helper(nums, 1, n - 1));
    }

    int helper(vector<int>& nums, int l, int r) {
        int sec_last = 0, last = 0, cur = 0;
        for(int i = l; i <= r; i++) {
            cur = max(last, nums[i] + sec_last);
            sec_last = last;
            last = cur;
        }
        return cur;
    }
};

File: mid/Dynamic Programming/198. House Robber II (Space-Opt DP).py
# https://leetcode.com/problems/house-robber-ii/solutions/4768883/97-1-approach-1-o-n-python-c-step-by-step-explanation/
# Time complexity : O(n)
# Space complexity : O(1)
# Space-Optimzed DP

class Solution:
    def rob(self, nums: List[int]) -> int:
        # 1. Robbing the first house and skipping the last house.
        # 2. Robbing the last house and skipping the first house.
        return max(nums[0], self.helper(nums[1:]), self.helper(nums[:-1]))

    def helper(self, nums):
        sec_last = last = cur = 0
        for i in nums:
            cur = max(last, sec_last + i)
            sec_last = last
            last = cur
        return cur